src/aliased_buffer.h

```diff
@@ -0,0 +1,193 @@
+
+#ifndef SRC_ALIASED_BUFFER_H_
+#define SRC_ALIASED_BUFFER_H_
+
+#include "v8.h"
+
+namespace node {
+
+/**
+ * This class encapsulates the technique of having a native buffer mapped to
+ * a JS object. Writes to the native buffer can happen efficiently without
+ * going through JS, and the data is then available to user's via the exposed
+ * JS object.
+ *
+ * While this technique is computationaly efficient, it is effectively a
+ * write to JS program state w/out going through the standard
+ * (monitored) API. Thus any VM capabilities to detect the modification are
+ * circumvented.
+ *
+ * The encapsulation herein provides a placeholder where such writes can be
+ * observed. Any notification APIs will be left as a future exercise.
+ */
+template <class NativeT, class V8T>
+class AliasedBuffer {
+ public:
+  AliasedBuffer(v8::Isolate* isolate, const size_t count) :
+    isolate_(isolate),
+    count_(count),
+    byte_offset_(0),
+    freeBuffer_(true) {
+    const v8::HandleScope handle_scope(this->isolate_);
+
+    const int sizeInBytes = sizeof(NativeT) * count;
+
+    // allocate native buffer
+    this->buffer_ = reinterpret_cast<NativeT*>(
+      calloc(count, sizeof(NativeT)));
+
+    // allocate v8 ArrayBuffer
+    v8::Local<v8::ArrayBuffer> ab = v8::ArrayBuffer::New(
+      this->isolate_, this->buffer_, sizeInBytes);
+
+    // allocate v8 TypedArray
+    v8::Local<V8T> jsArray = V8T::New(ab, this->byte_offset_, count);
+    this->jsArray_ = v8::Global<V8T>(isolate, jsArray);
+  }
+
+  /**
+   * Create an AliasedBuffer over a sub-region of another aliased buffer.
+   * The two will share a v8::ArrayBuffer instance &
+   * a native buffer, but will each read/write to different sections of the
+   * native buffer.
+   *
+   *  Note that byte_offset must by word-aligned.
+   */
+  AliasedBuffer(
+    v8::Isolate* isolate,
+    const size_t byte_offset,
+    const size_t count,
+    const AliasedBuffer<uint8_t,
+    v8::Uint8Array>& backingBuffer) :
+    isolate_(isolate),
+    count_(count),
+    byte_offset_(byte_offset),
+    freeBuffer_(false) {
+    const v8::HandleScope handle_scope(this->isolate_);
+
+    this->buffer_ = reinterpret_cast<NativeT*>(
+      const_cast<uint8_t*>(backingBuffer.GetNativeBuffer() + byte_offset));
+    v8::Local<v8::ArrayBuffer> ab = backingBuffer.GetArrayBuffer();
+    v8::Local<V8T> jsArray = V8T::New(ab, byte_offset, count);
+    this->jsArray_ = v8::Global<V8T>(isolate, jsArray);
+  }
+
+  AliasedBuffer(const AliasedBuffer& that) :
+    isolate_(that.isolate_),
+    count_(that.count_),
+    byte_offset_(that.byte_offset_),
+    buffer_(that.buffer_),
+    freeBuffer_(false) {
+    this->jsArray_ = v8::Global<V8T>(that.isolate_, that.GetJSArray());
+  }
+
+  ~AliasedBuffer() {
+    if (this->freeBuffer_ && this->buffer_ != NULL) {
+      free(this->buffer_);
+    }
+    // TODO(@mike-kaufman):  what do we do about freeing the
+    // v8::Global<ArrayBuffer> & the v8::Global<U>
```

Comment:
not clear to me if I need to do anything to manage lifecycle of a `v8::Global`?  Do I need to call `reset()`?

Reply:
That I'm unsure about. @addaleax or @bnoordhuis may know.

Reply:
A `Global<>` does need to be reset.

Reply:
fixed. calling `.Reset()` on the v8::Global.